Implement late feedback for Bluecurrent

Floris272 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This pr adresses the changes requested by @MartinHjelmare on the [initial pr](https://github.com/home-assistant/core/pull/82483).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Floris272 committed:
Apply changes

Floris272 committed:
Fix MockClient

MartinHjelmare commented on tests/components/blue_current/test_init.py:
```diff
@@ -46,176 +48,162 @@ async def test_load_unload_entry(hass: HomeAssistant) -> None:
-                "actual_kwh": 10,
-            }
+    mock_client = MagicMock(spec=Client)
+    connector = Connector(hass, config_entry, mock_client)
```
Don't use any integration details, like the connector. This test should just test that when the client callbacks with new data, an entity in the state machine should be updated with new state as expected.

The client callback is available as a mock call argument when patching the client, something like `mock_client_class.return_value.start_loop.call_args[0][0]`.

Floris272 replied:
Should i not test the connector itself? That is what test_on_data, test_start_loop and test_reconnect_* are for. Testing if the entity state is set correctly is done in test_sensor.py.

MartinHjelmare replied:
No, we should not test integration details directly. This allows us to to refactor the integration without changing the tests.

Floris272 replied:
Alright, so i think i should remove those tests in test_init.py right? (Since the entity state is tested in test_sensor.py).

MartinHjelmare replied:
Yes, if we already test it in the sensor tests we don't need to test it again here.

MartinHjelmare commented on tests/components/blue_current/test_sensor.py:
```diff
@@ -85,69 +84,106 @@
+    entity_registry = er.async_get(hass)
+
+    sensors = er.async_entries_for_config_entry(entity_registry, "uuid")
+    assert len(charge_point_status.keys()) + len(
```
We can remove `.keys()` when taking the length of a dict.
```suggestion
    assert len(charge_point_status) + len(
```

Floris272 committed:
Apply feedback

Floris272 committed:
Remove connector tests

Floris272 committed:
Change MockClient to inhert MagicMock

MartinHjelmare commented:
Looks like reconnecting isn't tested now.

Floris272 committed:
Add reconnect tests and refactor mock client

Floris272 commented:
> Looks like reconnecting isn't tested now.

Changed the reconnect tests to not call the connector directly. 

An issue i had was that reconnecting is now done in the background task so i could not use `async_fire_time_changed`. I fixed it by using an extra Event. I also feel like there is a better way to mock the methods than how it's done now. Could you take a look at it?

MartinHjelmare commented on tests/components/blue_current/__init__.py:
```diff
@@ -1,42 +1,121 @@
+    client.receiver = receiver
+    client.loop_start_task.set()
+
+    while True:
```
I'd use an `asyncio.Future` to wait on instead. We can set the result of the future from outside when we want the loop to finish or set an exception on the future if we want the loop to raise an exception.

MartinHjelmare commented on tests/components/blue_current/__init__.py:
```diff
@@ -1,42 +1,121 @@
+
+async def connect(client, token: str):
+    """Set the connected event."""
+    if client.connect_exception:
```
I'd just set a different side effect, ie the exception to raise, on the mock connect method instead.

Floris272 committed:
Refactor mock exception throwing

MartinHjelmare commented on tests/components/blue_current/__init__.py:
```diff
@@ -19,35 +19,21 @@
-
-
-async def start_loop(client, receiver):
+async def start_loop(hass: HomeAssistant, client, receiver):
```
I'd consider nesting this function inside the `create_client` function so we don't need to pass parameters that are not really part of the client method signature.

We can then create the future outside this function.

MartinHjelmare commented on tests/components/blue_current/__init__.py:
```diff
@@ -19,35 +19,21 @@
+    client.started_loop.set()
+    client.started_loop.clear()
+
+    client.loop_future = hass.loop.create_future()
```
I'd define a pytest fixture that creates the future and returns it. Use that fixture in the other fixtures or tests as needed.

We also then don't need to store the future on the client where it doesn't really belong.

MartinHjelmare reviewed (approved):
I think this looks ok. I've added a couple of comments for consideration for the future.

Floris272 committed:
Add future_fixture

Floris272 committed:
Move mocked client methods into create_client_mock

MartinHjelmare commented on tests/components/blue_current/__init__.py:
```diff
@@ -1,52 +1,94 @@
+) -> MagicMock:
+    """Create a mock of the bluecurrent-api Client."""
+    client_mock = MagicMock(spec=Client)
+    client_mock.started_loop = Event()
```
The `started_loop` event could also be created in a fixture similar as the future.

Floris272 replied:
I thought about it but the event is always needed. I could also remove it from mock_client and return it from create_client and init_integration to access it in the test instead.

MartinHjelmare replied:
I think it's more clear to have it separate from the mock client since it's not part of a real client.

Using functions or fixtures depends on the use case and test needs. You can mix and match that as you like.

MartinHjelmare commented on tests/components/blue_current/__init__.py:
```diff
@@ -1,52 +1,94 @@
+        client_mock.started_loop.set()
+        client_mock.started_loop.clear()
+
+        if loop_future and not loop_future.done():
```
Why is the future optional?

Floris272 replied:
Because the sensor tests don't need it. I could still pass the fixture but isn't this cleaner?

MartinHjelmare replied:
Don't we always want to make sure the background task is alive to test realistically?

Floris272 replied:
Moved the future creation to `init_integration` and the future is no longer optional. 

One thing is that to make it more realistic, the future should be recreated every time `start_loop` is called. Right now it just keeps throwing the exception set in a test. I don't really know how to do that and keep it accessible (in tests) other and creating a class that encapsulates the Future or something?

MartinHjelmare replied:
I'd define a pytest fixture that returns a dict or a dataclass that can store the future. Create the future in the `start_loop` side effect and then store the future in the dict or dataclass. Then the tests can access the future via the fixture.

Floris272 committed:
Remove fixture and separate event from mock_client

MartinHjelmare commented on tests/components/blue_current/__init__.py:
```diff
@@ -1,52 +1,97 @@
+    charge_point: dict | None = None,
+    status: dict | None = None,
+    grid: dict | None = None,
+) -> MagicMock:
```
Please adjust the return value typing to the new return value.

Floris272 committed:
Add FutureContainer to store the loop_future

MartinHjelmare commented:
Are we ready to merge?

Floris272 commented:
> Are we ready to merge?

Yes, thanks for your feedback!

MartinHjelmare commented:
When you want a review and are ready, you should click the "Ready for review" button. I did it for you this time.
